[Tablet Orders] Order list UI updates

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Partially closes #12149

## Description
This PR addresses most of the UI issues from the design review regarding the Order list. We have left outside of this PR those changes related to semantic colours, since all might need to be reviewed as per p1709628907203209/1708465836.162389-slack-C0354HSNUJH . This will be prioritized and tackled separately.

## Changes
* Adds a divider line to the top of the `OrderListTableViewController`

<img width="664" alt="Screenshot 2024-02-29 at 19 19 18" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/b8a048e1-0d39-4347-8d56-07b59007ad17">

* Adds divider lines to the selected cell in the order list, it would seem these were overwritten by the background colour

<img width="638" alt="Screenshot 2024-02-29 at 19 22 42" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/c789416b-a6e3-482c-b492-690f3c0a0f51">

* Updates the Empty Orders screen with a new image (we do not tackle the background colour update, nor removing the navigation bar)

![simulator_screenshot_88F19ADC-1F69-4724-859F-D4606402A23D](https://github.com/woocommerce/woocommerce-ios/assets/3812076/b1ff343a-d6c8-4123-8be9-ab60efedfbb2)

## Testing instructions
* With `sideBySideViewForOrderForm` enabled, on a tablet or large iPhone.
* Go to Orders. Observe that the dividers appear in the Order list to the left, both at the top, as well as between all Order cells.
* On the same device, on a freshly created store (or a store with no orders), go to Orders. Observe the new "shopping bags" image and  the "No Order Selected" text below it. (PD: This is something we could improve in the future, the "Waiting for your first Order" text should be more prominent in this case)

iamgabrielma committed:
Add divider to tableview’s top

iamgabrielma committed:
Update selected cell color

iamgabrielma committed:
Assure top and bottom borders to order tableview cells

iamgabrielma committed:
Move highlighted cell config to own method

iamgabrielma committed:
Add internal note

iamgabrielma committed:
Lint fix

iamgabrielma committed:
Revert selected cell color update

The current Woo colors might be out of date across the board, so we’re reverting the specific change on this commit. More details in the PR description.

iamgabrielma committed:
Update EmptyStateViewController with new image

iamgabrielma committed:
Lint fix: empty enum arguments violation

iamgabrielma commented on WooCommerce/Classes/ViewRelated/ReusableViews/EmptyStateViewController/EmptyStateViewController.swift:
```diff
@@ -439,20 +449,23 @@ extension EmptyStateViewController {
         fileprivate var message: NSAttributedString {
             switch self {
+            case .simpleTextWithDescription:
+                return NSAttributedString(string: "")
```
We cannot break here without retuning something, and making the message an optional `NSAttributedString` involves a bunch of changes everywhere we use the `message`. Returning an empty attributed string that is not being used seemed to fit our case without complicating it too much.

joshheald replied:
Consider mentioning this in a comment to save someone else refactoring unneccesarily

iamgabrielma replied:
👍  Added: https://github.com/woocommerce/woocommerce-ios/pull/12214/commits/abf24bdaf5fe5d56f3e248d642a61830f6143796

iamgabrielma committed:
Rename case for clarity

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift:
```diff
@@ -147,4 +149,17 @@ private extension OrderTableViewCell {
+        topBorder.backgroundColor = UIColor.divider.cgColor
+        layer.addSublayer(topBorder)
+        layer.addSublayer(bottomBorder)
+    }
```
These dividers don't quite look right, e.g. they're full width, and in this screenshot, one is placed over the default divider:
![cropped double divider](https://github.com/woocommerce/woocommerce-ios/assets/2472348/90f939c7-371a-4019-bf56-66b8a311e300)

Given that only the selected cell dividers are missing, I think it's better to leave the existing behaviour than apply this change. There's probably a better way to do it...

joshheald replied:
The fix should probably be in `UITableViewCell.updateDefaultBackgroundConfiguration(using:style:)` so we get it everywhere...

iamgabrielma replied:
> The fix should probably be in UITableViewCell.updateDefaultBackgroundConfiguration(using:style:) so we get it everywhere...

Yeah. I experimented with different approaches and the simplest one seems to be using the `.stroke...` properties inside the cell configuration. Here's the change: https://github.com/woocommerce/woocommerce-ios/pull/12214/commits/ffed3e920a4ea83949b52fb1ccd6cc342d4a3d03

I wasn't able to find a good way to apply this change only to top and bottom: 
* If we use CALayer we have to also play with removing them after a cell stops being selected/highlighted, which I wasn't able to do without crashing when populating the cells. This also adds additional unexpected lines around the first order and we'll have to add additional handling most likely by overriding `layoutSubviews`
* Using the [customView](https://developer.apple.com/documentation/uikit/uibackgroundconfiguration/3600322-customview) property feels a bit overkill to add these borders, at least for the moment. If we happen to redo the cells across the app this could make more sense.

The downside of the approach is that we can see the lateral border a bit wider when a cell is selected, but I think it makes sense and overall is an improvement:
![simulator_screenshot_39C4B052-902D-4FA2-B141-D366C88AF3E6](https://github.com/woocommerce/woocommerce-ios/assets/3812076/b14f886b-9d7b-48bd-9a78-09465a8f8434)

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -330,6 +330,13 @@ private extension OrderListViewController {
+        let borderWidth = 0.5
+        topBorder.frame = CGRect(x: 0, y: 0, width: tableView.frame.size.width, height: borderWidth)
+        topBorder.backgroundColor = UIColor.divider.cgColor
+        tableView.layer.addSublayer(topBorder)
```
This one's in the wrong place – when you pull to refresh, the divider comes with it. It should be at the bottom of the header, I think...
![IMG_5593BBA5DCD0-1](https://github.com/woocommerce/woocommerce-ios/assets/2472348/125ea4f8-c85b-419b-b0cc-52130a2f108d)

iamgabrielma replied:
Updated! https://github.com/woocommerce/woocommerce-ios/pull/12214/commits/03f5cb645e9051cf6854bb035ab9d7f315f50062

joshheald commented:
>     * Updates the Empty Orders screen with a new image (we do not tackle the background colour update, nor removing the navigation bar)
> 
> 
> ![simulator_screenshot_88F19ADC-1F69-4724-859F-D4606402A23D](https://private-user-images.githubusercontent.com/3812076/310364878-b1ff343a-d6c8-4123-8be9-ab60efedfbb2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3NDM0MTgsIm5iZiI6MTcwOTc0MzExOCwicGF0aCI6Ii8zODEyMDc2LzMxMDM2NDg3OC1iMWZmMzQzYS1kNmM4LTQxMjMtOGJlOS1hYjYwZWZlZGZiYjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDZUMTYzODM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjBkZWFiNGYyNWY3NjM4OTYzYzJmYWVhODE1OTAxMDcxY2NiNmIzYjQyMzk4MDZjZDM0NzcwYWU2ODFhOGQ5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.IuFjzT8EhD_4XYm1iRcOqiw_8DO1Fpy9K3Isna1rEjg)

The two types of empty state for the list/details panes look a bit rubbish next to each other. Let's try to update both, at least as a follow up issue...
@joe-keenan if you happen to have images for the ones on the left, that'd be appreciated, but it's super low priority and we can muddle through if not.

![IMG_8F6A4EE8FA99-1](https://github.com/woocommerce/woocommerce-ios/assets/2472348/afa89ee2-f975-413b-b2e1-2199fbad126b)

joshheald reviewed (commented):
Not sure this one's ready, we can keep working on it after release...

joe-keenan commented:
Hey @joshheald, just acknowledging that I’ve seen this and will get back to you with something for the empty state next week. Let me know if you need it sooner.

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma commented:
> something for the empty state

👋  @joe-keenan did you had the chance to look into this? Don't worry if not, as Josh mentioned is very low priority, I'm just catching up with this issue so I wanted to be sure before I update the existing ones.

iamgabrielma committed:
Merge branch 'trunk' into issue/12149-order-list-ui-updates

iamgabrielma committed:
Add comment

iamgabrielma committed:
Move border to bottom of filtered orders header

iamgabrielma committed:
Remove duplication for backgroundColor set

iamgabrielma committed:
remove unnecessary self

iamgabrielma committed:
Apply border to cell default configuration

iamgabrielma commented:
@joshheald this is ready for review when you have the time:

**Adressed:**
- Moved the border/separator line from the top or the orders to the bottom of the search component, so it looks fine when we _pull to refresh_
- Moved the border/separator lines for highlighted cells to the `UITableViewCell.updateDefaultBackgroundConfiguration(using:style:)` method
- Added internal comments

**Not addressed:**
- Cohesive images when there are two types of empty state for the list/details. Created a follow-up here: https://github.com/woocommerce/woocommerce-ios/issues/12625

joe-keenan commented:
> > something for the empty state
> 
> 👋 @joe-keenan did you had the chance to look into this? Don't worry if not, as Josh mentioned is very low priority, I'm just catching up with this issue so I wanted to be sure before I update the existing ones.

Hey @iamgabrielma, sorry about the delay. I was about to say I hadn’t done anything for this, but then I realised I had made a similar empty state for Analytics a while back.

I reused the format and illustration to fit the orders list. I also feel like the Clear Filters CTA overwhelms the error message, so I changed it to a secondary button style.

Let me know if you have any issues with it!

6kkCwI9VOEYcidp6UT5bln-fi-1725_29745

![image](https://github.com/woocommerce/woocommerce-ios/assets/95003933/e5d98500-9e73-4a73-ad64-10b77a36f6ed)

iamgabrielma commented:
> I reused the format and illustration to fit the orders list. I also feel like the Clear Filters CTA overwhelms the error message, so I changed it to a secondary button style.

Thanks! Looks good! I tried to update this today but requires a bit more work than I expected. Eg:

- We reuse this view's configuration in different places, eg: changing the colour of the button forces us to change it in other "empty" views that are be unrelated to orders, or contain just a link rather than a button.
- Moving order of the image and the text forces us to create an entire new view, since what we have cannot be reused.
- It also has to be updated when the modal-on-modal returns an empty search.

With the current focus I'll move this latest update to the issue that is already open on https://github.com/woocommerce/woocommerce-ios/issues/12625 for empty states, so we can merge the order improvements for the time being. I haven't made any further changes since the latest update, so the addressed vs not addressed items remains the same 👍

joe-keenan commented:
> With the current focus I'll move this latest update to the issue that is already open on https://github.com/woocommerce/woocommerce-ios/issues/12625 for empty states, so we can merge the order improvements for the time being.

That’s fine. I also would like a more cohesive approach to error states. Thanks for creating a separate issue for it.

joshheald reviewed (approved):
Thanks for the changes, they look good to me 👍
